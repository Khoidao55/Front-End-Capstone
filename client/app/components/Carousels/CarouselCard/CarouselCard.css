#card {
  position: relative;
  width: 300px;
  height: 575px;
  background-color: white;
  border: 2px solid black;
}

#product-image {
  border-bottom: 1px solid black;
}

.descriptive {
  padding-left: 10px;
}

#product-name {
  margin: 10px;
}

#price {
  color: gray;
}

#action {
  position: absolute;
  right: 5px;
  top: 5px;
  z-index: 2;
}
