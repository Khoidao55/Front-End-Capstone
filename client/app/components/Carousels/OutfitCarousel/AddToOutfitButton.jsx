import React from 'react';
import store from './storeAPI';

const AddToOutfit = ({ productInfo }) => (
  <button
    type="button"
    id="add-to-outfit-btn"
    className="card"
    onClick={() => {
      console.log('in button productInfo:', productInfo);
      if (productInfo.id !== 0 && !store.has(productInfo.id)) {
        store.save(productInfo.id, productInfo);
      }
    }}
  >
    <i className="fas fa-plus-square" />
    <br />
    Add To Outfit
  </button>
);

export default AddToOutfit;
