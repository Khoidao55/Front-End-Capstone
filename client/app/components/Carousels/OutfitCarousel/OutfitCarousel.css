#outfit-carousel {
  position: relative;
  overflow: hidden;
  margin-bottom: 100px;
  min-height: 37.5rem;
  border-top: 2px solid var(--Nodolla-purple);
  border-bottom: 2px solid var(--Nodolla-purple);
  display: grid;
  grid-template-columns: 10% 80% 10%;
  place-items: center;
  background-color: thistle;
}

#outfit-carousel-title {
  margin: 20px 20px 0 10px;
  display: inline-block;
}

#outfit-card-container {
  width: 100%;
  grid-column: 2;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  padding: 5px 0;
  flex-wrap: nowrap;
  overflow: hidden;
}


#add-to-outfit-btn {
  outline: none;
  width: 16rem;
  height: 580px;
  margin-right: 100px;
  margin-left: 5px;
  border-top-right-radius: 10px;
  border-top-left-radius: 10px;
  border: 3px solid var(--Nodolla-purple);
  color: var(--Nodolla-purple);
  font-size: 1.5rem;
  font-weight: 700;
  line-height: 36px;
  background-color: rgba(242, 242, 242, .6);
  box-shadow: 5px 7px 0 #aaaaaa;
  transition: box-shadow .5s;
  -webkit-transition: box-shadow .5s;
}

#add-to-outfit-btn:hover {
  cursor: pointer;
  box-shadow: 1px 1px 7px rgb(90, 90, 90);
  transition: box-shadow .5s;
  -webkit-transition: box-shadow .5s;
}

.fa-plus-square {
  font-size: 36px;
  font-weight: 500;
}