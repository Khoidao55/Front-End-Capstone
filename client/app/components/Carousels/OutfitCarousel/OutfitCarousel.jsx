import React from 'react';

const OutfitCarousel = () => (

);

export default OutfitCarousel;
