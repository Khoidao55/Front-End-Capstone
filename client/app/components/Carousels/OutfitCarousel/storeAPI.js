/* ************************************ */
/*    This is a simple API to expand    */
/*   my interaction with localStorage   */
/* ************************************ */

const subscribers = new Set();

/* ******* */
/*   get   */
/* ******* */

// Input: a key to search for in localStorage
// Output: A value stored with this key as its index or null.

// Description:
//              Searchs localStorage for an item with the supplied key as its index.
//              Returns the value of that item if it is found.
//              Returns null if no item with this key exists in localStorage
const get = (key) => (
  JSON.parse(localStorage.getItem(key))
);

/* ********** */
/*   getAll   */
/* ********** */

// Input: N/A
// Output: An array containing all data currently stored in localStorage.

// Description:
//              Retreives all items stored in localStorage as an array of objects.
//              Returns an empty array if no items are currently stored in localStorage.
const getAll = () => {
  const { length, ...relevantData } = localStorage;

  return Object.keys(relevantData).map((key) => (
    get(key)
  ));
};

/* ******* */
/*   has   */
/* ******* */

// Input: a key to search for in localStorage
// Output: boolean

// Description:
//              Searches localStorage for an item with the supplied key.
//              Returns true if it is found, otherwise returns false.
const has = (key) => {
  if (get(key)) {
    return true;
  }
  return false;
};

const subscribe = (cb) => {
  subscribers.add(cb);
};

const unsubscribe = (cb) => {
  subscribers.delete(cb);
};

const publish = (subs, state) => {
  subs.forEach((sub) => {
    sub(state);
  });
};

const doPublish = () => publish(subscribers, getAll());

/* ******** */
/*   save   */
/* ******** */

// Input: A unique key and a value to save to the localStorage.
// Output: N/A

// Description:
//              Saves a new item to localStorage with the supplied unique key
//              as its index and the value as its value.
//              The value is stringifyed to be parsed later by const get().
//              If an item with the supplied key already exists in localStorage, throws an error
//              and doesn't save the item.
const save = (key, value) => {
  if (!has(key)) {
    localStorage.setItem(key, JSON.stringify(value));
    doPublish();
  } else {
    console.error('An item with this key already exists in localStorage.\n'
      + 'If you would like to update the item with this key, use const update()');
  }
};

/* ********** */
/*   remove   */
/* ********** */

// Input: A key for an item to remove from localStorage
// Output: The value of the removed item or null

// Description:
//              removes an item in localStorage with the supplied key as its index.
//              Returns the value of the removed item, or null if no item with that key
//              exists in localStorage.
const remove = (key) => {
  const toRemove = get(key);
  localStorage.removeItem(key);
  doPublish();
  return toRemove;
};

/* ************* */
/*   removeAll   */
/* ************* */

// Input: N/A
// Output: N/A

// Description:
//              removes all data from localStorage.
const removeAll = () => {
  localStorage.clear();
  doPublish();
};

/* ******** */
/*   size   */
/* ******** */

// Input: N/A
// Output: The number of items stored in localStorage

// Description:
//              Returns the number of items stored in localStorage
const size = () => (
  localStorage.length
);

/* ********** */
/*   update   */
/* ********** */

// Input: a key that already exists in localStorage and a new value
// Output: N/A

// Description:
//              Updates an existing item with that has supplied key as its index
//              with a new value.
//              If no item with this key exists in localStorage, throws an error
//              and doesn't update the item.
const update = (key, newValue) => {
  if (has(key)) {
    localStorage.setItem(key, JSON.stringify(newValue));
    doPublish();
  } else {
    console.error('There is no item with this key in localStorage.\n'
      + 'If you would like to save new item, use const save()');
  }
};

/* ********************** */
/*   bestHTTPStatusCode   */
/* ********************** */

// Input: N/A
// Output: The best HTTP status code

// Description:
//              Self Explainatory.
const bestHTTPStatusCode = () => (
  '418: I\'m a little teapot.'
);

export default {
  save,
  getAll,
  remove,
  removeAll,
  update,
  has,
  size,
  subscribe,
  unsubscribe,
  bestHTTPStatusCode,
};
