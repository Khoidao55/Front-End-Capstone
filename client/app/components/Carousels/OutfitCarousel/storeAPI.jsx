/* ************************************ */
/*    This is a simple API to expand    */
/*   my interaction with localStorage   */
/* ************************************ */

const store = {};

/* ******** */
/*   save   */
/* ******** */

// Input: a key value pair to save to the localStorage.
// The input is stringifyed to be parsed on a get.
// Output: N/A
store.save = (key, value) => {
  localStorage.setItem(key, JSON.stringify(value));
};

/* ******* */
/*   get   */
/* ******* */

// Input: a key to search for in localStorage
// Output: A value stored with this key as its index
store.get = (key) => (
  JSON.parse(localStorage.getItem(key))
);

/* ********** */
/*   getAll   */
/* ********** */

// Input: N/A
// Output: An array containing all data currently stored in localStorage
store.getAll = () => {
  const { length, ...relevantData } = localStorage;
  const dataToReturn = [];
  const keys = Object.keys(relevantData);

  keys.forEach((key) => {
    dataToReturn.push(store.get(key));
  });

  return dataToReturn;
};

/* ******* */
/*   has   */
/* ******* */

// Input: a key to search for in localStorage
// Output: true if an item with this key exists, false if not
store.has = (key) => {
  if (store.get(key)) {
    return true;
  }
  return false;
};

export default store;
