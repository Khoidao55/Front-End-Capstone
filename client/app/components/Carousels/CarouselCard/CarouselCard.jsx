/* eslint-disable import/extensions */
/* eslint-disable jsx-a11y/control-has-associated-label */
import React from 'react';
import propTypes from 'proptypes';
import StarRating from '../../StarRating/StarRating.jsx';
import './CarouselCard.css';

const CarouselCard = ({
  id,
  productImage = 'https://watertownbusinesscoalition.com/assets/images/no_image_available.jpeg',
  category,
  name,
  originalPrice,
  salePrice,
  reviewScore,
  handleActionButton,
  handleRedirect,
  carouselType,
  features,
}) => {
  let displayPrice;

  if (salePrice !== null) {
    displayPrice = (
      <>
        <span className="descriptive old-price">{`$${originalPrice}`}</span>
        <span className="descriptive display-price">{`$${salePrice}`}</span>
      </>
    );
  } else {
    displayPrice = <span className="descriptive" id="display-price">{`$${originalPrice}`}</span>;
  }

  return (
    <div className={`card ${carouselType}-card`} onClick={() => { handleRedirect(id) }} >
      <button
        aria-label="Action Button"
        className="card-action-btn"
        type="button"
        onClick={(e) => { e.stopPropagation(); handleActionButton(id, name, features); }}
      >
        <i className={
          carouselType === 'related'
            ? 'related-action-btn fas fa-star'
            : 'outfit-action-btn far fa-times-circle'
        }
        />
      </button>
      <div className="card-image-container">
        {productImage
          ? <img className="card-image" src={productImage} alt={name ? name : 'Product Name'} />
          : (
            <img
              className="card-no-image"
              src="https://watertownbusinesscoalition.com/assets/images/no_image_available.jpeg"
              alt={name ? name : 'Product Name'}
            />
          )}
      </div>
      <div className="descriptive-container">
        <span className="descriptive category">{category ? category : 'Category'}</span>
        <h3 className="descriptive product-name">{name ? name : 'Product Name'}</h3>
        {displayPrice}
        <StarRating reviewScore={reviewScore} setMargin="10px 0 0 10px" />
      </div>
    </div>
  );
};

CarouselCard.propTypes = {
  id: propTypes.number.isRequired,
  productImage: propTypes.string,
  category: propTypes.string.isRequired,
  name: propTypes.string.isRequired,
  originalPrice: propTypes.string.isRequired,
  salePrice: propTypes.string,
  reviewScore: propTypes.number,
  handleActionButton: propTypes.func.isRequired,
  handleRedirect: propTypes.func.isRequired,
  carouselType: propTypes.string.isRequired,
};

export default CarouselCard;
