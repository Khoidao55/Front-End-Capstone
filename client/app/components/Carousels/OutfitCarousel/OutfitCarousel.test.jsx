import React from 'react';
import { shallow } from 'enzyme';
import CarouselCard from '../CarouselCard/CarouselCard.jsx';
import OutfitCarousel from './OutfitCarousel.jsx';
import AddToOutfitButton from './AddToOutfitButton.jsx';

const dummyProduct = {
  id: 1234,
  productImage: 'https://bsaber.com/wp-content/uploads/2020/01/a2040f737cc603cf561c54624bde21715b7cbf52.jpg',
  category: 'meme',
  name: 'pretty cool bananas',
  originalPrice: '1234567',
  salePrice: '123',
  stars: '3.5',
};

const handleRedirect = () => { console.log('Redirect!'); };

describe('Outfit Carousel', () => {
  test('OutfitCarousel component renders correctly', () => {
    const wrapper = shallow(<OutfitCarousel productInfo={dummyProduct} handleRedirect={handleRedirect} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('AddToOutfitButton component renders correctly', () => {
    const wrapper = shallow(<AddToOutfitButton productInfo={dummyProduct} />);

    expect(wrapper).toMatchSnapshot();
  });

});
