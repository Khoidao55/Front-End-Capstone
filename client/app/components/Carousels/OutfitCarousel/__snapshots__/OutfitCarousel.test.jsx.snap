// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outfit Carousel AddToOutfitButton component renders correctly 1`] = `ShallowWrapper {}`;

exports[`Outfit Carousel OutfitCarousel component renders correctly 1`] = `ShallowWrapper {}`;
